larpandoradlcontent/LArVertex/DlVertexingAlgorithm.cc

```diff
@@ -0,0 +1,932 @@
+            xMin = (xVtx + xJitter) - m_regionSize;
+            xMax = (xVtx + xJitter) + m_regionSize;
+            zMin = (zVtx + zJitter) - m_regionSize;
+            zMax = (zVtx + zJitter) + m_regionSize;
+        }
+
+        LArMvaHelper::MvaFeatureVector featureVector;
+        featureVector.emplace_back(static_cast<double>(nuance));
+        featureVector.emplace_back(static_cast<double>(nVertices));
+        featureVector.emplace_back(xVtx);
+        featureVector.emplace_back(zVtx);
+
+        for (const CaloHit *pCaloHit : *pCaloHitList)
+        {
+            const float x{pCaloHit->GetPositionVector().GetX()}, z{pCaloHit->GetPositionVector().GetZ()}, adc{pCaloHit->GetInputEnergy()};
+            // If on a refinement pass, drop hits outside the region of interest
+            if (m_pass > 1 && (x < xMin || x > xMax || z < zMin || z > zMax))
+                continue;
+            featureVector.emplace_back(static_cast<double>(x));
+            featureVector.emplace_back(static_cast<double>(z));
+            featureVector.emplace_back(static_cast<double>(adc));
+            ++nHits;
+        }
+        featureVector.insert(featureVector.begin() + 2, static_cast<double>(nHits));
+        // Only write out the feature vector if there were enough hits in the (possibly perturbed) region of interest
+        if (nHits > 10)
+            LArMvaHelper::ProduceTrainingExample(trainingFilename, true, featureVector);
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//-----------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode DlVertexingAlgorithm::Infer()
+{
+    if (m_pass == 1)
+        ++m_event;
+    CartesianPointVector vertexCandidatesU, vertexCandidatesV, vertexCandidatesW;
+    for (const std::string listName : m_caloHitListNames)
+    {
+        const CaloHitList *pCaloHitList{nullptr};
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetList(*this, listName, pCaloHitList));
+
+        HitType view{pCaloHitList->front()->GetHitType()};
+        const bool isU{view == TPC_VIEW_U}, isV{view == TPC_VIEW_V}, isW{view == TPC_VIEW_W};
+        if (!isU && !isV && !isW)
+            return STATUS_CODE_NOT_ALLOWED;
+
+        LArDLHelper::TorchInput input;
+        PixelVector pixelVector;
+        this->MakeNetworkInputFromHits(*pCaloHitList, input, pixelVector);
+
+        // Run the input through the trained model
+        LArDLHelper::TorchInputVector inputs;
+        inputs.push_back(input);
+        LArDLHelper::TorchOutput output;
+        if (isU)
+            LArDLHelper::Forward(m_modelU, inputs, output);
+        else if (isV)
+            LArDLHelper::Forward(m_modelV, inputs, output);
+        else
+            LArDLHelper::Forward(m_modelW, inputs, output);
+
+        double thresholds[]{0., 0.00275, 0.00825, 0.01925, 0.03575, 0.05775, 0.08525, 0.12375, 0.15125, 0.20625, 0.26125, 0.31625, 0.37125,
```

Comment:
Do these need to be hardcoded?  If so, could they be stored in another file so the maintenance is (a little bit) easier.

Reply:
It ought to be possible to make these configurable and read them from XML

Reply:
Is all of the training produced in a single job?  Or are multiple jobs used where the seed is the same?

Reply:
Good point, it's multiple jobs, so there will be some seed sharing going on here